mono/metadata/icall.c

```diff
@@ -6891,53 +6891,83 @@ ves_icall_System_Environment_GetCommandLineArgs (MonoError *error)
 	return result;
 }
 
-#ifndef HOST_WIN32
-static MonoArray *
-mono_icall_get_environment_variable_names (MonoError *error)
-{
-	MonoArray *names;
-	MonoDomain *domain;
-	MonoString *str;
-	gchar **e, **parts;
-	int n;
+MonoArrayHandle
+ves_icall_System_Environment_GetEnvironmentVariableNames (MonoError *error)
+{
+	MonoDomain * const domain = mono_domain_get ();
+	gsize n = 0;
+	MonoStringHandle str = MONO_HANDLE_NEW (MonoString, NULL);
+
+#ifdef HOST_WIN32
+	WCHAR * const env_strings = GetEnvironmentStringsW ();
+
+	if (env_strings) {
+		WCHAR const *env_string = env_strings;
+		while (*env_string) {
+			/* weird case that MS seems to skip (per drive letter current working directory) */
+			if (*env_string != '=')
+				n++;
+			while (*env_string)
+				env_string++;
+			env_string++;
+		}
+	}
 
-	error_init (error);
-	n = 0;
-	for (e = environ; *e != 0; ++ e)
-		++ n;
+	MonoArrayHandle names = mono_array_new_handle (domain, mono_defaults.string_class, n, error);
+	return_val_if_nok (error, NULL_HANDLE_ARRAY);
 
-	domain = mono_domain_get ();
-	names = mono_array_new_checked (domain, mono_defaults.string_class, n, error);
-	return_val_if_nok (error, NULL);
+	if (env_strings) {
+		n = 0;
+		WCHAR const *env_string = env_strings;
+		while (*env_string) {
+			/* weird case that MS seems to skip (per drive letter current working directory) */
+			if (*env_string != '=') {
+				WCHAR const * const equal_str = wcschr (env_string, '=');
+				g_assert (equal_str);
+				mono_string_new_utf16_to_handle (str, domain, env_string, (gsize)(equal_str - env_string), error);
```

Comment:
There is no need to introduce this new concept of `_to_handle`, expect if you come up with clear benchmark results pointing otherwise. Is it a backed-by-data problem, or is it something you just came up with? _I understand it will avoid allocating handles, no need to lecture on that_

For sake of getting this PR in, revert to using `mono_string_new_utf16_handle`, and submit another PR with this specific change, and we can discuss it there.

Reply:
Assigning through out parameters is an old concept.

`to_array` is a new concept I just came up with and is not wearing well as I do more conversions, i.e. in custom-attr.c, which lead me to revisit it here.

I've had to repeatedly split up loops into multiple functions, making the code messier and slower.  It took me a while to realize the better approach.
The idiom of `_to_handle`, or better name,  `_assign`, avoids that repeatedly.